server/service/src/login.rs

```diff
@@ -184,12 +206,21 @@ impl LoginService {
             }
         };
 
+        Ok(user_info)
+    }
+
+    pub fn update_user(
+        service_ctx: &ServiceContext,
+        username: &str,
+        password: &str,
+        user_info: LoginUserInfoV4,
+    ) -> Result<(), UpdateUserError> {
         // convert user_info to internal format
         let user = UserAccountRow {
             id: user_info.user.id,
             username: username.to_string(),
             hashed_password: UserAccountService::hash_password(&password)
-                .map_err(|err| FetchUserError::InternalError(format!("{:?}", err)))?,
+                .map_err(|e| UpdateUserError::CreateUserAccountError(e))?,
```

Comment:
InternalError is probably sufficient here, the bcrypt error shouldn't really be happening. Otherwise CreateUserAccountError -> UpdateUserAccountError or HashError?

Reply:
> InternalError is probably sufficient here

I think it is, but I like having all of the error types specified in error enum (can see what are all of the possible errors without digest all of the code).

> Otherwise CreateUserAccountError -> UpdateUserAccountError or HashError

It looks like has_password was returning CreateUserAccountError vs explicit BcryptError (I think you were pretty strong on using map_err and just returning the error that's in the helper method vs generalising errors), so I've changed this not for. hash_password to just return BcryptError:  bf58ecd6d44174f8d1b454199b1e2d710b361da8

Reply:
Fine with it even though this error is so unlikely that upper layers shouldn't really bother about it.